internal/pkg/agent/application/upgrade/manual_rollback.go

```diff
@@ -0,0 +1,258 @@
+
+	var selectedRollbackVersionedHome string
+
+	err = withTakeOverWatcher(ctx, log, topDir, watcherHelper, func() error {
+		// read the upgrade marker
+		updateMarker, err = LoadMarker(paths.DataFrom(topDir))
+		if err != nil {
+			return fmt.Errorf("loading marker: %w", err)
+		}
+
+		if updateMarker == nil {
+			return ErrNilUpdateMarker
+		}
+
+		if len(updateMarker.RollbacksAvailable) == 0 {
+			return ErrNoRollbacksAvailable
+		}
+
+		for versionedHome, rollback := range updateMarker.RollbacksAvailable {
+			if rollback.Version == version && now.Before(rollback.ValidUntil) {
+				selectedRollbackVersionedHome = versionedHome
+				break
+			}
+		}
+		if selectedRollbackVersionedHome == "" {
+			return fmt.Errorf("version %q not listed among the available rollbacks: %w", version, ErrNoRollbacksAvailable)
+		}
+		return nil
+	})
+
+	if err != nil {
+		// Invoke watcher again (now that we released the watcher applocks)
+		_, invokeWatcherErr := watcherHelper.InvokeWatcher(log, watcherExecutable)
+		if invokeWatcherErr != nil {
+			return "", "", errors.Join(err, fmt.Errorf("invoking watcher: %w", invokeWatcherErr))
+		}
+		return "", "", err
+	}
+
+	return watcherExecutable, selectedRollbackVersionedHome, nil
+}
+
+func withTakeOverWatcher(ctx context.Context, log *logger.Logger, topDir string, watcherHelper WatcherHelper, f func() error) error {
+	watcherApplock, err := watcherHelper.TakeOverWatcher(ctx, log, topDir)
+	if err != nil {
+		return fmt.Errorf("taking over watcher processes: %w", err)
+	}
+	defer func(watcherApplock *filelock.AppLocker) {
+		releaseWatcherAppLockerErr := watcherApplock.Unlock()
+		if releaseWatcherAppLockerErr != nil {
+			log.Warnw("error releasing watcher applock", "error", releaseWatcherAppLockerErr)
+		}
+	}(watcherApplock)
+
+	return f()
+}
+
+func extractAgentInstallsFromMarker(updateMarker *UpdateMarker) (previous agentInstall, current agentInstall, err error) {
+	previousParsedVersion, err := version.ParseVersion(updateMarker.PrevVersion)
+	if err != nil {
+		return previous, current, fmt.Errorf("parsing previous version %q: %w", updateMarker.PrevVersion, err)
+	}
+	previous = agentInstall{
+		parsedVersion: previousParsedVersion,
+		version:       updateMarker.PrevVersion,
+		hash:          updateMarker.PrevHash,
+		versionedHome: updateMarker.PrevVersionedHome,
+	}
+
+	currentParsedVersion, err := version.ParseVersion(updateMarker.Version)
+	if err != nil {
+		return previous, current, fmt.Errorf("parsing current version %q: %w", updateMarker.Version, err)
+	}
+	current = agentInstall{
+		parsedVersion: currentParsedVersion,
+		version:       updateMarker.Version,
+		hash:          updateMarker.Hash,
+		versionedHome: updateMarker.VersionedHome,
+	}
+
+	return previous, current, nil
+}
+
+func getAvailableRollbacks(rollbackWindow time.Duration, now time.Time, currentVersion string, parsedCurrentVersion *version.ParsedSemVer, currentVersionedHome string, currentHash string) map[string]TTLMarker {
```

Comment:
Doesn't look like `currentHash` is being used?

Reply:
please have a look at 1cd2407e1b